removed mirror store

EmanekaT opened the pull request:
I removed mirror_store, and changed all keys in dynamodb_state_store from shelvekey to regular string.
Tested with make test. I kept shelvestore there since I don't think we necessarily need to remove it.

keymone commented on tron/serialize/runstate/dynamodb_state_store.py:
```diff
@@ -119,37 +112,37 @@ def save(self, key_value_pairs) -> None:
             self.alert(str(e))
 
-    def __setitem__(self, key: ShelveKey, val: bytes) -> None:
+    def __setitem__(self, key: str, val: bytes) -> None:
```
does the caller know that they must no longer send `ShelveKey` objects here?

EmanekaT replied:
statemanager is the caller and it calls build_key to obtain keys? So it shouldn't have knowledge about ShelveKey at the first place. It was ShelveKey because the caller was mirror_state_store.

qui commented on tron/serialize/runstate/dynamodb_state_store.py:
```diff
@@ -30,18 +28,13 @@ def restore(self, keys) -> dict:
-                if str(key) in items:
-                    translated_items[key] = items[str(key)]
         except Exception as e:
             self.alert(str(e))
```
Please log this too. Also, what should happen if Tron is unable to restore the state of a job when starting up? It looks like we would return empty items, which would cause Tron to start overwriting the state for all jobs with empty state. We definitely don't want that, try to handle it so the impact is as small as possible.

This didn't affect us before when we were relying on the state from bdb, but it will once we stop using bdb.

EmanekaT replied:
I think restore should not and cannot restore anything if restore failed. If that ever happens, shelvestore could be used as a backup? Or we can manually restore dynamodb states from backup.

keymone commented on tron/serialize/runstate/dynamodb_state_store.py:
```diff
@@ -119,37 +112,37 @@ def save(self, key_value_pairs) -> None:
         except Exception as e:
             self.alert(str(e))
```
since we're using batch_writer we're safe(ish?) with regard to partial saves: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb.html#DynamoDB.Table.batch_writer

> ... the batch writer will also automatically handle any unprocessed items and resend them as needed.

it's still possible batch_writer will fail even after retrying write operations so would be useful to catch this case somehow.

keymone replied:
of course the right choice here is to use transact_write_items instead of batch_writer: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb.html#DynamoDB.Client.transact_write_items

EmanekaT replied:
transact_write_item is a good idea. But I also have to add a retry alogrithm to it. In the worst case, I think we have to lose some data. I can set up scheduled backup to recover though.

EmanekaT commented:
Changes:
* updated __setitem__ to use transact_write_items. Writing items with more than 10 partitions is still not atomic because transact_write_items could only handle up to 10 items at once. We only have one job with around 13 partitions across all clusters right now so this is probably not a big deal.
* updated __get_items_ so that it fails after retrying 10 times and throws an exception 
*  transact_write_items/batch_get_items do not retry after failure, so I implemented retrying myself. 
* updated alerts to include different error messages for different failures. It still only emails me until it is thoroughly tested.
* Upgraded requirements using requirements_tools because transact_write_items is only available in the latest version 
* using new version of boto3 with moto might cause an error on travis. I might need to resolve it.

EmanekaT commented:
I will add more tests for cases like failed restore later.

EmanekaT commented on tests/serialize/runstate/dynamodb_state_store_test.py:
```diff
@@ -1,25 +1,59 @@
 
 
-filename = os.path.join(tempfile.mkdtemp(), 'state')
+def mock_transact_write_items(self):
```
It is added as an instance method of client class. That's why it's named as self. I think name it self makes more sense, and I can add a comment explain what self is in the context.
And I can add comments to explain how it works. The code is from a Japanese site so linking it doesn't make much sense.

qui replied:
Comments sound good, thanks.

EmanekaT commented on tests/serialize/runstate/dynamodb_state_store_test.py:
```diff
@@ -63,6 +97,8 @@ def large_object():
 
 @pytest.mark.usefixtures("store", "small_object", "large_object")
 class TestDynamoDBStateStore:
+    @mock.patch('moto.dynamodb2.responses.DynamoHandler.transact_write_items',
```
I tried to patch store with transact_write_item but it doesn't work.
